SDWebImage/SDWebImageManager.m

```diff
@@ -292,27 +299,26 @@ - (void)saveImageToCache:(nullable UIImage *)image forURL:(nullable NSURL *)url
 }
 
 - (void)cancelAll {
-    @synchronized (self.runningOperations) {
-        NSArray<SDWebImageCombinedOperation *> *copiedOperations = [self.runningOperations copy];
-        [copiedOperations makeObjectsPerformSelector:@selector(cancel)];
-        [self.runningOperations removeObjectsInArray:copiedOperations];
-    }
+    LOCK(self.runningOperationsLock);
+    NSArray<SDWebImageCombinedOperation *> *copiedOperations = [self.runningOperations copy];
+    UNLOCK(self.runningOperationsLock);
+    [copiedOperations makeObjectsPerformSelector:@selector(cancel)]; // This will call `safelyRemoveOperationFromRunning:` and remove from the array
 }
 
 - (BOOL)isRunning {
     BOOL isRunning = NO;
-    @synchronized (self.runningOperations) {
-        isRunning = (self.runningOperations.count > 0);
-    }
+    LOCK(self.runningOperationsLock);
+    isRunning = (self.runningOperations.count > 0);
+    UNLOCK(self.runningOperationsLock);
     return isRunning;
 }
 
 - (void)safelyRemoveOperationFromRunning:(nullable SDWebImageCombinedOperation*)operation {
-    @synchronized (self.runningOperations) {
-        if (operation) {
-            [self.runningOperations removeObject:operation];
-        }
+    LOCK(self.runningOperationsLock);
+    if (operation) {
```

Comment:
This line of code can be a little better. Check if operation is nil before, only it's not nil we need to lock the runningOperations.

```objective-c
if (!operation) {
    return;
}
LOCK(self.runningOperationsLock);
[self.runningOperations removeObject:operation];
UNLOCK(self.runningOperationsLock);
```

Reply:
It's indeed better :)